File: packages/unity-game/Assets/Scripts/Core/Random/DeterministicRng.cs
using System;

namespace GLD.Core.Random
{
    public sealed class DeterministicRng
    {
        uint _state;

        public DeterministicRng(uint seed)
        {
            _state = seed;
        }

        public uint NextUint32()
        {
            unchecked
            {
                _state = 1664525u * _state + 1013904223u;
                return _state;
            }
        }

        public float NextFloat01()
        {
            return NextUint32() / 4294967296f;
        }

        public int NextInt(int maxExclusive)
        {
            if (maxExclusive <= 0)
                throw new ArgumentOutOfRangeException(nameof(maxExclusive), "maxExclusive must be positive.");
            return (int)Math.Floor(NextFloat01() * maxExclusive);
        }

        public float NextRange(float min, float max)
        {
            if (max < min)
                throw new ArgumentOutOfRangeException(nameof(max), "max must be greater than or equal to min.");
            return min + (max - min) * NextFloat01();
        }
    }
}

File: packages/shared/src/random/deterministic-rng.ts
export class DeterministicRng {
	private state: number;

	constructor(seed: number) {
		this.state = seed >>> 0;
	}

	nextUint32(): number {
		this.state = (Math.imul(1664525, this.state) + 1013904223) >>> 0;
		return this.state;
	}

	nextFloat01(): number {
		return this.nextUint32() / 0x100000000;
	}

	nextInt(maxExclusive: number): number {
		if (!Number.isInteger(maxExclusive) || maxExclusive <= 0) {
			throw new RangeError('maxExclusive must be a positive integer');
		}
		return Math.floor(this.nextFloat01() * maxExclusive);
	}

	nextRange(min: number, max: number): number {
		if (max < min) {
			throw new RangeError('max must be greater than or equal to min');
		}
		return min + (max - min) * this.nextFloat01();
	}
}
